fix: consistent colors for packages

Shaharking opened the pull request:
### PR Description

This PR addresses issue [#2564](https://github.com/vercel/turbo/issues/2564) tagged as a good first issue.

**Changes:**

-   Used hashing on the package name to ensure consistent color assignment.
-   Expanded the color set for better distinction.

**Additional Notes:**

-   I encountered challenges running all the integration tests and was unable to figure out how to execute the `turborepo-tests`. It might be beneficial to provide a more detailed explanation of the testing process in the `CONTRIBUTING.md` file to assist future contributors.

### Testing Instructions
- The tests today already cover everything

chris-olszewski commented on crates/turborepo-ui/src/color_selector.rs:
```diff
@@ -1,21 +1,20 @@
-            Style::new().yellow(),
-            Style::new().blue(),
-        ]
+        let colors: [Style; u8::MAX as usize] =
+            core::array::from_fn(|index| Style::new().color256(index as u8));
+
+        colors
```
I don't believe we want to use all of the available colors as:
 - many of them aren't especially nice to look at
 - many of the colors are very similar e.g. in my tests I got 4 very similar shades of red

<img width="384" alt="Screenshot 2024-08-14 at 1 32 23 PM" src="https://github.com/user-attachments/assets/604076b6-278d-4d48-8202-fc25172a4f9c">

chris-olszewski replied:
I'm still opposed to using all of the colors here for the reasons listed above. 

We can expand our list in a separate PR, but we need to make sure the colors selected are easy to read in most terminals.

chris-olszewski reviewed (changes requested):
Thank you for the PR! Do not worry about the integration tests, those will run on CI.

I think what we want is to keep the existing colors, but to hand them out in a consistent way while also making use of all the available colors i.e. not assigning 2 tasks the same color if there's still colors without any tasks.

To be frank, the current setup is overly complicated as it happens during the graph walk which is non-deterministic. I would suggest:
 - Prepopulating the color cache with all keys [when we start the walk](https://github.com/vercel/turborepo/blob/main/crates/turborepo-lib/src/task_graph/visitor.rs#L159). This will look something like:
 ```
for task in engine.tasks().sorted() {
    self.color_cache.color_for_key(&task.to_string());
}
```
 - Then also switch so we select colors with the task id as well [here](https://github.com/vercel/turborepo/blob/main/crates/turborepo-lib/src/task_graph/visitor.rs#L670)

Shaharking commented:
Got it @chris-olszewski.

I see that you suggested using the sorted task name to choose a color and caching it in a hashmap to ensure consistent color assignment. A straightforward approach would be to use an index and increment it each time we need a color for a new task.

Since the tasks are ordered by name, this method will generally guarantee the same color for a monorepo that doesn't change much.

However, if the order changes—like when adding or removing a package—there’s a good chance the colors will no longer match.

We could take a more complex approach to ensure consistent colors even when new packages are added. This could involve hashing the task name and using the result modulo the number of unused colors.

(hash % unused_colors.len())

The number of available colors decreases each time a color is assigned. Once all colors are used, we start again with the full color collection.

I think that for new projects that add lots of packages and change frequently, this feature might not be very beneficial. While approach 2 is a bit more complicated, it's still straightforward enough. I’m inclined to go with this approach—let me know your thoughts.

Shaharking commented:
Hey @chris-olszewski,

I've updated the PR to include the following:

-   **Prepopulating the color cache**: All keys are now prepopulated when we start the walk, as suggested. 
```
for task in engine.tasks().sorted() {
    self.color_cache.color_for_key(&task.to_string());
}
```
    
-   **Selecting colors using the task ID**: I’ve also switched the color selection to use the task ID, as recommended [here](https://github.com/vercel/turborepo/blob/main/crates/turborepo-lib/src/task_graph/visitor.rs#L670).

Regarding the issue of similar colors: What do you suggest? We could limit the palette to more distinct colors, but this might result in more frequent repetition.

chris-olszewski commented on crates/turborepo-ui/src/color_selector.rs:
```diff
@@ -65,13 +74,40 @@ impl ColorSelectorInner {
+        hasher.finish()
+    }
+
+    fn get_color_id_by_key(&mut self, key: &str) -> usize {
```
This implementation suffers from the same issue as the existing one where adding/removing tasks will result in a changed color for the task when there are hash collisions. If both `a#build` and `b#build` get the same initial `color_id`, then removing `a#build` will result in `b#build` changing colors.

Given we're targeting such a small hash (5 at the moment, 256 if we allow for all colors) hash collisions are fairly likely.

chris-olszewski reviewed (changes requested):
Just the `visitor.rs` changes in this PR are a huge improvement:
 - color order is no longer tied to task execution order
 - color is no longer tied to task hashes which change frequently

I would prefer to merge these improvements as they will make life better for many developers. As you mentioned there are still issues with tasks being added/removed from the graph, but that can be addressed in future PRs.

If you would move the changes to `color_selector.rs` to a new PR, then I would happily merge the remaining changes.

Shaharking commented:
Sure @chris-olszewski,
Here I created a seprate pr
https://github.com/vercel/turborepo/pull/9023

Shaharking commented:
@chris-olszewski 
I think we can increase the palette to 14 distinct colors. You can check out these references for more information:

    https://superuser.com/questions/783656/whats-the-deal-with-terminal-colors
    https://askubuntu.com/questions/1305637/what-do-the-colors-in-a-terminal-color-scheme-correspond-do
    https://github.com/mbadolato/iTerm2-Color-Schemes

I excluded black because most users use it as their background color.

This change should reduce color collisions between packages, and having 14 colors, as opposed to 5, should make it easier to visually distinguish issues.

Shaharking commented:
Hey @chris-olszewski, is increasing the number of colors worth pursuing, or do you prefer to skip it? I'm unsure whether to continue testing different colors due to the lack of response.

ijjk commented:
Allow CI Workflow Run

- [ ] approve CI run for commit: 46f49c4563f77a36c99c5a2561707681838983e0

Note: this should only be enabled once the PR is ready to go and can only be enabled by a maintainer

ijjk commented:
Allow CI Workflow Run

- [ ] approve CI run for commit: 46f49c4563f77a36c99c5a2561707681838983e0

Note: this should only be enabled once the PR is ready to go and can only be enabled by a maintainer

chris-olszewski commented:
@Shaharking Apologies for dropping communication. The team is currently not interested in expanding the possible color list.
